refactor(import): Introduce ImportService interface and unify method naming solving issue #1115

HuzaifaOmar opened the pull request:
# Pull Requests Requirements

-  [x] The PR title includes a brief description of the work done, including the Issue number if applicable.  
- [ ] The PR includes a video showing the changes for the work done.  
- [x] The PR title follows conventional commit label standards.  
- [x] The changes conform to the OpenElis Global x3 Styleguide and design documentation.  
- [x] The changes include tests or are validated by existing tests.  
- [x] I have read and agree to the Contributing Guidelines of this project.  

## **Summary**  
This PR introduces an `ImportService` interface to improve code modularity and maintainability. It unifies method names across different import services, replacing `importOrganizationList()` and `importPractitionerList()` with `importList()`. The changes help streamline the import logic and remove redundancy in `ImportController`.  

## Screenshots

[Add relevant screenshots here if applicable]

## Related Issue
#1115 

## Other

[Add any additional information or notes here]

HuzaifaOmar committed:
Refactor Import Logic: Introduce ImportService Interface and Unify Method Naming

HuzaifaOmar committed:
Merge branch 'develop' into importcontroller-duplicate-logic-1115

HuzaifaOmar committed:
refactor: apply formatter

HuzaifaOmar committed:
Merge branch 'importcontroller-duplicate-logic-1115' of github.com:HuzaifaOmar/OpenELIS-Global-2 into importcontroller-duplicate-logic-1115

Agaba-derrick commented:
@HuzaifaOmar  have you tried to run mvn spotless:apply  to format your changes  ?

does the redanducy in the import controller affect the import services in  a way that would require us this change ?

HuzaifaOmar commented:
@Agaba-derrick 
**Regarding formatting:** Yes, I ran the command `mvn spotless:apply` in the commit titled "refactor: apply formatter".

**Regarding the necessity of these changes:** Since I’m still new to the organization, I picked up [issue #1115](https://github.com/I-TECH-UW/OpenELIS-Global-2/issues/1115) as an introductory task. As I understand, this refactor was requested to make the code more future-proof and encourage clean code, especially if additional services are added in the future. Let me know if there's anything I might have overlooked!.

mozzy11 committed:
Merge branch 'develop' into importcontroller-duplicate-logic-1115

mozzy11 committed:
Merge branch 'develop' into importcontroller-duplicate-logic-1115

mozzy11 commented on src/main/java/org/openelisglobal/organization/service/OrganizationImportService.java:
```diff
@@ -3,8 +3,9 @@
-public interface OrganizationImportService {
+public interface OrganizationImportService extends ImportService {
 
-    void importOrganizationList() throws FhirLocalPersistingException, FhirGeneralException, IOException;
```
Its a good practice not to modify methods in a service Interface

HuzaifaOmar replied:
I just renamed the method name to have a common interface

mozzy11 replied:
its a not a good practice to modify the method signature in an interface

mozzy11 commented on src/main/java/org/openelisglobal/provider/service/ProviderImportService.java:
```diff
@@ -4,7 +4,7 @@
+public interface ProviderImportService extends ImportService {
 
-    void importPractitionerList() throws FhirLocalPersistingException, FhirGeneralException, IOException;
+    void importList() throws FhirLocalPersistingException, FhirGeneralException, IOException;
```
Its a good practice not to modify methods in a service Interface

mozzy11 commented on src/main/java/org/openelisglobal/provider/service/ImportService.java:
```diff
@@ -0,0 +1,9 @@
+package org.openelisglobal.provider.service;
+
```
This new service looks redundant

mozzy11 commented on src/main/java/org/openelisglobal/admin/controller/ImportController.java:
```diff
@@ -14,19 +15,41 @@
+
+    private void importDataFromFhir(ResourceType resourceType)
+            throws FhirLocalPersistingException, FhirGeneralException, IOException {
+        ImportService importService;
```
you can get rid of this redundant new service

HuzaifaOmar commented:
Ok, I will modify the code to remove the interface I made and just undo the method naming changes

HuzaifaOmar committed:
Refactor: Remove ImportService interface

Removed the `ImportService` interface and adjusted related classes to directly implement the required methods.  This simplifies the codebase and removes unnecessary abstraction.

HuzaifaOmar commented:
Hey @mozzy11, I've resolved the comments you left on my code. When you have a moment, could you take another look? Thanks!

HuzaifaOmar committed:
Merge branch 'develop' into importcontroller-duplicate-logic-1115

mherman22 commented on src/main/java/org/openelisglobal/admin/controller/ImportController.java:
```diff
@@ -14,19 +14,38 @@
+            break;
+        default:
+            // Handle invalid resource type
+            throw new UnsupportedOperationException("Unsupported resource type");
```
Since all possible cases are covered (and the default should never be reached), this exception might be unnecessary unless more resource types will be added later.

HuzaifaOmar replied:
While the default case may seem redundant, I think it's a good defensive programming that aligns with the issue's focus on future maintainability and potential new resource types. It makes the code's future-proofing intentions explicit. If you still find it useless I'll remove it.

mherman22 reviewed (approved):
Thanks @HuzaifaOmar , one slight comment otherwise this looks good to me 💯

mherman22 commented:
Feel free to jump on any other issue as we wait for @mozzy11's last input on this.
do you need help with this https://github.com/I-TECH-UW/OpenELIS-Global-2/issues/1062?

HuzaifaOmar commented:
@mherman22  I did work on issue #1062, but I noticed that `SpringContext.getBean()` is used extensively throughout the codebase. I’d like to propose working on this issue partially, as I may not be able to refactor all instances of `SpringContext.getBean()` at once. Would that be acceptable?

mherman22 commented:
> @mherman22 I did work on issue #1062, but I noticed that `SpringContext.getBean()` is used extensively throughout the codebase. I’d like to propose working on this issue partially, as I may not be able to refactor all instances of `SpringContext.getBean()` at once. Would that be acceptable?

That's okay.

mozzy11 commented:
Thanks @HuzaifaOmar , and thanks @mherman22 for the review
